File: 2022/day/9/main.py
from dataclasses import dataclass


@dataclass
class Snek:
    headx: int
    heady: int
    tailx: int
    taily: int
    
    @property
    def head(self):
        return (self.heady, self.headx)

    @property
    def tail(self):
        return (self.taily, self.tailx)
    
    def move(self, motion):
        stepdir, _steplen = motion.split(" ", maxsplit=2)
        steplen = int(_steplen)
        
        hx, hy = self.headx, self.heady
        tx, ty = self.tailx, self.taily
        
        match stepdir:
            case "U": 
                dx, dy = 0, -1
            case "D":
                dx, dy = 0, 1
            case "L":
                dx, dy = -1, 0
            case "R":
                dx, dy = 1, 0

        for _ in range(steplen):
            hx += dx
            hy += dy
            
            if abs(hx - tx) > 1:
                tx += dx
                ty = hy

            if abs(hy - ty) > 1:
                tx = hx
                ty += dy
                
            assert 0 <= abs(hx - tx) <= 1
            assert 0 <= abs(hy - ty) <= 1
            yield Snek(hx, hy, tx, ty)


assert next(Snek(0, 0, 0, 0).move("R 1")) == Snek(1, 0, 0, 0), "R 1"
assert next(next(Snek(0, 0, 0, 0).move("R 1")).move("L 1")) == Snek(0, 0, 0, 0), "R 1 . L 1 == id"
assert next(Snek(1, 0, 0, 0).move("R 1")) == Snek(2, 0, 1, 0)
assert next(Snek(1, 0, 0, 0).move("L 1")) == Snek(0, 0, 0, 0)
assert next(Snek(2, 1, 1, 0).move("D 1")) == Snek(2, 2, 2, 1)
assert next(Snek(2, 1, 1, 0).move("L 1")) == Snek(1, 1, 1, 0)


with open("input.txt") as f:
    motions = [ln.strip() for ln in f]

snek = Snek(0, 0, 0, 0)
seen = set()
seen.add(snek.tail)

for motion in motions:
    for snek in snek.move(motion):
        seen.add(snek.tail)
        
# Part 1
print(len(seen))

File: 2022/day/9/Main.hs
{-# LANGUAGE ImportQualifiedPost, ViewPatterns #-}
import Control.Arrow (second)
import Control.Monad
import Data.Set qualified as Set
import Data.Traversable


type Rope = [(Int, Int)]

-- Work out where the tail goes based on the (updated) head
adjust :: (Int, Int) -> (Int, Int) -> (Int, Int)
adjust (hy, hx) (ty, tx)
    | abs dy <= 1 && abs dx <= 1 = (ty, tx)
    | abs dy >= 2 && abs dx >= 2 = (ty', tx')
    | abs dy >= 2 = (ty', hx)
    | abs dx >= 2 = (hy, tx')
    | otherwise = (tx, ty)
    where
        dy = hy - ty
        dx = hx - tx
        ty'
            | ty < hy = hy - 1
            | otherwise = hy + 1
        tx'
            | tx < hx = hx - 1
            | otherwise = hx + 1

translate (dy, dx) (y, x) = (y + dy, x + dx)

-- Apply motion to rope; returns the "trail".
stepN :: String -> Rope -> [Rope]
stepN ('U':' ':(read -> n)) = take n . tail . steps (-1, 0)
stepN ('D':' ':(read -> n)) = take n . tail . steps (1, 0)
stepN ('L':' ':(read -> n)) = take n . tail . steps (0, -1)
stepN ('R':' ':(read -> n)) = take n . tail . steps (0, 1)

-- Repeatedly move the rope in one direction
steps :: (Int, Int) -> Rope -> [Rope]
steps dir = iterate (\(x:xs') -> scanl adjust (translate dir x) xs')

-- Apply all motions to the rope; returns the final rope and the
-- the whole trail.
simulate :: Rope -> [String] -> (Rope, [Rope])
simulate rope = second concat . mapAccumL (\xs motion -> let xs' = stepN motion xs in (last xs', xs')) rope


main = do
    input <- readFile "input.txt"
    
    -- Part 1
    let rope1 = [(0, 0), (0, 0)]
        (_, trail1) = simulate rope1 (lines input)
        tails1 = map last trail1
    print $ length $ Set.fromList tails1

    -- Part 2
    let rope2 = replicate 10 (0, 0)
        (_, trail2) = simulate rope2 (lines input)
        tails2 = map last trail2
    print $ length $ Set.fromList tails2
